core/src/main/scala/org/apache/spark/rdd/RDDScope.scala

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.rdd
+
+import java.util.concurrent.atomic.AtomicInteger
+import org.apache.spark.SparkContext
+
+/**
+ * A collection of utility methods to construct a hierarchical representation of RDD scopes.
+ * An RDD scope tracks the series of operations that created a given RDD.
+ */
+private[spark] object RDDScope {
+
+  // Symbol for delimiting each level of the hierarchy
+  // e.g. grandparent;parent;child
+  val SCOPE_NESTING_DELIMITER = ";"
```

Comment:
It seems a bit awkward to be defining our own string serialization format for scopes. This is actually a public API (right?) because we will be writing this out to the event logs and reading it back. What if instead you just had:

1. A programmatic representation of a scope.
2. A JSON serializer for that representation.

The JSON output would likely be a list of enclosing scopes.